src/iter/panic_fuse.rs

```diff
@@ -0,0 +1,338 @@
+use super::plumbing::*;
+use super::*;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::thread;
+
+/// `PanicFuse` is an adaptor that wraps an iterator with a fuse in case
+/// of panics, to halt all threads as soon as possible.
+///
+/// This struct is created by the [`panic_fuse()`] method on [`ParallelIterator`]
+///
+/// [`panic_fuse()`]: trait.ParallelIterator.html#method.panic_fuse
+/// [`ParallelIterator`]: trait.ParallelIterator.html
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+#[derive(Debug, Clone)]
+pub struct PanicFuse<I: ParallelIterator> {
+    base: I,
+}
+
+/// Helper that sets a bool to `true` if dropped while unwinding.
+#[derive(Clone)]
+struct Fuse<'a>(&'a AtomicBool);
+
+impl<'a> Drop for Fuse<'a> {
+    #[inline]
+    fn drop(&mut self) {
+        if thread::panicking() {
+            self.0.store(true, Ordering::Relaxed);
+        }
+    }
+}
+
+impl<'a> Fuse<'a> {
+    #[inline]
+    fn panicked(&self) -> bool {
+        self.0.load(Ordering::Relaxed)
+    }
+}
+
+/// Create a new `PanicFuse` iterator.
+///
+/// NB: a free fn because it is NOT part of the end-user API.
+pub fn new<I>(base: I) -> PanicFuse<I>
+where
+    I: ParallelIterator,
+{
+    PanicFuse { base: base }
+}
+
+impl<I> ParallelIterator for PanicFuse<I>
+where
+    I: ParallelIterator,
+{
+    type Item = I::Item;
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: UnindexedConsumer<Self::Item>,
+    {
+        let panicked = AtomicBool::new(false);
+        let consumer1 = PanicFuseConsumer {
+            base: consumer,
+            fuse: Fuse(&panicked),
+        };
+        self.base.drive_unindexed(consumer1)
+    }
+
+    fn opt_len(&self) -> Option<usize> {
+        self.base.opt_len()
+    }
+}
+
+impl<I> IndexedParallelIterator for PanicFuse<I>
+where
+    I: IndexedParallelIterator,
```

Comment:
I'm not sure we should really have this, nor `opt_len()`. It's _safe_, but consumers like `collect_vec()` are going to fail their own assertions if they come up short. If nothing else, that means we may have competing panics.

That's also a problem for the idea of implicitly fusing in the `bridge` functions.

Reply:
Interesting point. But this is true also for the original, right? i.e., if you do `par_iter().map(|x| panic!()).collect()`, it'll have the same problem, no?

Reply:
The difference in yours is that individual `CollectConsumer` splits will either be fully completed, or they'll be panic-unwound without getting a chance to `complete()` and notice that they're short. Same for the overall `Collect::complete()`, it will just be unwound.

But with `panic_fuse()`, when one consumer panics, the other remaining consumers will return normally, and then `CollectFolder::complete()` has a chance to notice the shortfall and raise a panic of its own.

I guess we could remove the "too few" assertion in `CollectFolder::complete()` and just leave it to the overall `Collect::complete()` to check the total. That last one should still be skipped by the fuse's panic unwinding.

Reply:
> I guess we could remove the "too few" assertion in `CollectFolder::complete()` and just leave it to the overall `Collect::complete()` to check the total. That last one should still be skipped by the fuse's panic unwinding.

I've implemented this, with new tests to ensure we get the panic we want.